File: testdata/fixtures/ruby/CWE-78.rb
# CWE-78 OS Command Injection in Ruby
# Ruby's shell APIs execute strings via /bin/sh when given a single argument.
# If that string is built from request parameters, an attacker can append
# arbitrary commands with ; or &&.

require 'open3'

def index
  user = params[:cmd]
  q = params["input"]
  ck = cookies[:cmd]
  body = request.body.read
  qs = request.query_string

  # ruleid: ruby-command-injection
  system(user)

  # ruleid: ruby-command-injection
  exec(q)

  # ruleid: ruby-command-injection
  Open3.popen3(ck)

  # ruleid: ruby-command-injection
  Kernel.system(body)

  # ok: ruby-command-injection
  system("ls -l /tmp")

  # ok: ruby-command-injection
  exec("echo hello")

  # ok: ruby-command-injection
  Kernel.system("fixed_command")
end

def extra
  extra_qs = request.query_string
  # ruleid: ruby-command-injection
  spawn(extra_qs)
  # ok: ruby-command-injection
  spawn("ls")
end

File: testdata/fixtures/typescript/CWE-78.ts
// CWE-78: OS Command Injection – user input fed directly to child_process
// execution sinks allows arbitrary OS command execution.
// Sources are req.query / params ; sinks are exec, execSync, spawn.

import * as child_process from "child_process";
import { Request } from "express";

function vuln(req: Request) {
  const cmd = req.query.get("cmd") as string;
  // ruleid: typescript-command-injection
  child_process.exec(cmd);
  const arg = req.params.get("arg") as string;
  // ruleid: typescript-command-injection
  child_process.execSync(arg);
  const body = req.body.get("body") as string;
  // ruleid: typescript-command-injection
  child_process.spawn(body, []);
  const q = location.search;
  // ruleid: typescript-command-injection
  child_process.exec(q);
}

function safe() {
  // ok: typescript-command-injection
  child_process.exec("ls -la");
  // ok: typescript-command-injection
  child_process.execSync("echo hi");
  // ok: typescript-command-injection
  child_process.spawn("ls", ["-la"]);
  // ok: typescript-command-injection
  child_process.execFile("ls", ["-la"]);
}
